graphql_compiler/schema_merging/rename_schema.py

```diff
@@ -0,0 +1,232 @@
+# Copyright 2019-present Kensho Technologies, LLC.
+from collections import namedtuple
+
+from graphql import build_ast_schema, parse
+from graphql.language.visitor import Visitor, visit
+
+from .utils import SchemaError, get_schema_data
+
+
+RenamedSchema = namedtuple(
+    'RenamedSchema', [
+        'schema_ast',  # type: Document, ast representing the renamed schema
+        'reverse_name_map',  # type: Dict[str, str], new type name to old type name
+        'reverse_root_field_map'  # type: Dict[str, str], new root field name to old root field
```

Comment:
Nit: What is a `root field`? I don't believe this is defined anywhere.

Reply:
By "root field" I meant "fields of the root type", aka fields of the query type. Where would be a good place to put this explanation? Here at the definition of RenamedSchema?

Reply:
What do you think about prohibiting renaming of the query type altogether? I don't really see a use case that would require it, and I don't think it should matter what it is named so long as it doesn't conflict with any other type name. In that case, we won't ever need this value, since we can just look it up from the schema object itself.

The more functionality we can cut out without losing the core value proposition, the faster and better we'll be able to implement the rest.

Reply:
The description is definitely not good here -- I meant it as in, for example, fields of `RootSchemaQuery`. So if we have `RootSchemaQuery { Human: Human }`, this part renames the first `Human`

Reply:
Ah I see, I definitely got confused. How about prohibiting divergences there as well, and enforcing that in the query type, the fields are named the same as the types they hold? Then I think we'd just be able to use the `reverse_name_map` directly.

Also, it might be good to mention in the comment next to `reverse_name_map` whether all types (even non-renamed ones) have entries, or types that aren't there are not renamed. I'd probably opt for the second, since we have some very large schemas.

Reply:
I was hesitant to enforce that the fields are named the same as the types, because GraphQL doesn't require this (and the basic GraphQL tutorial examples has query fields like `hero: Human`). Is this restriction enforced somewhere in graphql compiler?

Reply:
I definitely understand the hesitation, it's very reasonable. We don't currently enforce this at the moment, because the compiler originally did not concern itself with how the schema was built, and simply assumed that it would be given a reasonable schema to work with. However, I wouldn't be surprised if having field names that don't match the type name on the query type of the schema causes some unexpected bugs to pop up in the compiler. So I think we should start enforcing it, first here and then probably in the rest of the compiler codebase as well.

Reply:
> Also, it might be good to mention in the comment next to `reverse_name_map` whether all types (even non-renamed ones) have entries, or types that aren't there are not renamed. I'd probably opt for the second, since we have some very large schemas.

In order to check for conflicts, I need to keep track of all names, even ones that are not changed, during the renaming step. At the end of that I'm going to have a map with all types, including non-renamed ones. Would it be better to keep that full dict and put it into the output? Or process it afterwards by removing all entries that didn't get renamed?